File: linked-list/21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == nullptr && list2 == nullptr){
            return nullptr;
        }

        if(list1 == nullptr || list2 == nullptr){
            return list1 == nullptr ? list2 : list1;
        }

        ListNode* crrFirst = list1;
        ListNode* crrSecond = list2;
        ListNode* result = nullptr;

        if(crrFirst->val < crrSecond->val){
                result = crrFirst;
                crrFirst = crrFirst->next;
        }else{
                result = crrSecond;
                crrSecond = crrSecond->next;
        }

        ListNode* head = result;

        while(crrFirst != nullptr && crrSecond != nullptr){
            if(crrFirst->val < crrSecond->val){
                result->next = crrFirst;
                crrFirst = crrFirst->next;
            }else{
                result->next = crrSecond;
                crrSecond = crrSecond->next;
            }

            result = result->next;
        }

        if(crrFirst != nullptr){
            result->next = crrFirst;
        }

        if(crrSecond != nullptr){
            result->next = crrSecond;
        }

        return head;
    }
};

File: Linked-List/21-merge-two-sorted-lists/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if(list1 == null && list2 == null){
            return null;
        }

        if(list1 == null || list2 == null){
            return list1 != null ? list1 : list2;
        }

        ListNode crrFirst = list1;
        ListNode crrSecond = list2;

        ListNode result = null;

        if(crrFirst.val < crrSecond.val){
            result = crrFirst;
            crrFirst = crrFirst.next;
        }else{
            result = crrSecond;
            crrSecond = crrSecond.next;
        }

        ListNode head = result;

        while(crrFirst != null && crrSecond != null){
            if(crrFirst.val <= crrSecond.val){
                result.next = crrFirst;
                crrFirst = crrFirst.next;
            }else{
                result.next = crrSecond;
                crrSecond = crrSecond.next;
            }

            result = result.next;
        }

        if(crrFirst == null && crrSecond != null){
            result.next = crrSecond;
        }

        if(crrFirst != null && crrSecond == null){
            result.next = crrFirst;
        }

        return head;
    }
}